File: client/ios/App/Navigation/MobileRouteResolver.swift
import Foundation

// Marketing America Corp. Oleksandr Tishchenko
public enum MobileRouteResolver {
    public static func resolve(_ rawRoute: String?, query: [String: String] = [:]) -> MobileRoute? {
        let route = normalizeRoute(rawRoute)
        let segments = route.split(separator: "/").map(String.init)

        switch route {
        case "dashboard": return .dashboard
        case "more": return .more
        case "access/sign-in": return .accessSignIn
        case "access/register": return .accessRegister
        case "access/recovery/request": return .accessRecoveryRequest
        case "access/recovery/reset": return .accessRecoveryReset
        case "access/verification": return .accessVerification
        case "access/password": return .accessPassword
        case "access/sign-out": return .accessSignOut
        case "vendor": return .vendor
        case "vendor/page": return .vendorProfile
        case "vendor/summary": return .vendorSummary
        case "vendor/statement": return .vendorStatement
        case "vendor/payout": return .vendorPayout
        case "vendor/transaction": return .vendorTransaction
        case "vendor/attachment": return .vendorAttachment
        case "catalog": return .catalog
        case "catalog/browse": return .catalogBrowse
        case "catalog/search": return .catalogSearch(searchText: query["q"] ?? query["searchText"])
        case "message": return .message
        case "notification": return .notification
        case "attachment": return .attachment
        case "cart": return .cart
        case "cart/checkout": return .cartCheckout
        case "vendor/product": return .vendorProduct
        case "vendor/product/new": return .vendorProductNew
        case "vendor/order": return .vendorOrder
        case "vendor/order/new": return .vendorOrderNew
        case "vendor/project": return .vendorProject
        case "vendor/project/new": return .vendorProjectNew
        default:
            return resolveSegmentedRoute(segments)
        }
    }

    public static func resolve(_ link: MobileLink) -> MobileRoute? {
        resolve(link.route, query: link.query)
    }

    public static func isKnownRoute(_ rawRoute: String?) -> Bool {
        resolve(rawRoute) != nil
    }

    public static func isSignOutAction(action: String?, route rawRoute: String?) -> Bool {
        action == "access.sign_out" || resolve(rawRoute) == .accessSignOut
    }

    public static func isCurrentlyRenderable(_ rawRoute: String?) -> Bool {
        switch resolve(rawRoute) {
        case .dashboard,
             .more,
             .vendor,
             .vendorProfile,
             .vendorSummary,
             .vendorStatement,
             .vendorPayout,
             .vendorTransaction,
             .vendorAttachment,
             .vendorProduct,
             .vendorProductNew,
             .vendorProductDetail,
             .vendorOrder,
             .vendorOrderNew,
             .vendorOrderDetail,
             .vendorOrderShipment,
             .vendorOrderShipmentDetail,
             .vendorOrderTax,
             .vendorOrderTaxDetail,
             .vendorProject,
             .vendorProjectNew,
             .vendorProjectDetail,
             .message,
             .notification,
             .catalog,
             .attachment:
            return true
        default:
            return false
        }
    }

    public static func normalizeRoute(_ rawRoute: String?) -> String {
        let normalized = (rawRoute ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .split(separator: "/", omittingEmptySubsequences: true)
            .joined(separator: "/")

        return normalized == "vendor/profile" ? "vendor/page" : normalized
    }

    private static func resolveSegmentedRoute(_ segments: [String]) -> MobileRoute? {
        if segments.count == 3, segments[0] == "catalog", segments[1] == "node" {
            return .catalogNode(catalogNodeId: segments[2])
        }
        if segments.count == 3, segments[0] == "message", segments[1] == "thread" {
            return .messageThread(threadId: segments[2])
        }
        if segments.count == 2, segments[0] == "attachment" {
            return .attachmentDetail(attachmentId: segments[1])
        }
        if segments.count == 4, segments[0] == "cart", segments[1] == "checkout", segments[2] == "result" {
            return .cartCheckoutResult(checkoutId: segments[3])
        }
        if segments.count == 3, segments[0] == "vendor", segments[1] == "product" {
            return .vendorProductDetail(productId: segments[2])
        }
        if segments.count == 3, segments[0] == "vendor", segments[1] == "order" {
            return .vendorOrderDetail(orderId: segments[2])
        }
        if segments.count == 4, segments[0] == "vendor", segments[1] == "order", segments[3] == "shipment" {
            return .vendorOrderShipment(orderId: segments[2])
        }
        if segments.count == 5, segments[0] == "vendor", segments[1] == "order", segments[3] == "shipment" {
            return .vendorOrderShipmentDetail(orderId: segments[2], shipmentId: segments[4])
        }
        if segments.count == 4, segments[0] == "vendor", segments[1] == "order", segments[3] == "tax" {
            return .vendorOrderTax(orderId: segments[2])
        }
        if segments.count == 5, segments[0] == "vendor", segments[1] == "order", segments[3] == "tax" {
            return .vendorOrderTaxDetail(orderId: segments[2], taxId: segments[4])
        }
        if segments.count == 3, segments[0] == "vendor", segments[1] == "project" {
            return .vendorProjectDetail(projectId: segments[2])
        }
        return nil
    }
}

File: client/android/app/src/main/java/app/mobiling/client/navigation/MobileRouteResolver.kt
package app.mobiling.client.navigation

/**
 * Marketing America Corp. Oleksandr Tishchenko
 */
object MobileRouteResolver {
    fun resolve(rawRoute: String?, query: Map<String, String> = emptyMap()): MobileRoute? {
        val route = normalizeRoute(rawRoute)
        val segments = route.split('/').filter { it.isNotBlank() }

        return when {
            route == "dashboard" -> MobileRoute.Dashboard
            route == "more" -> MobileRoute.More
            route == "money" -> MobileRoute.Money
            route == "access/sign-in" -> MobileRoute.Access.SignIn
            route == "access/register" -> MobileRoute.Access.Register
            route == "access/recovery/request" -> MobileRoute.Access.RecoveryRequest
            route == "access/recovery/reset" -> MobileRoute.Access.RecoveryReset
            route == "access/verification" -> MobileRoute.Access.Verification
            route == "access/password" -> MobileRoute.Access.Password
            route == "access/sign-out" -> MobileRoute.Access.SignOut
            route == "wallet" -> MobileRoute.Wallet.Overview
            route == "wallet/transaction" -> MobileRoute.Wallet.Transaction
            route == "wallet/funding" -> MobileRoute.Wallet.Funding
            route == "wallet/withdrawal" -> MobileRoute.Wallet.Withdrawal
            segments.size == 3 && segments[0] == "wallet" && segments[1] == "withdrawal" -> MobileRoute.Wallet.WithdrawalDetail(segments[2])
            route == "vendor" -> MobileRoute.Vendor.Overview
            route == "vendor/page" -> MobileRoute.Vendor.Profile
            route == "vendor/summary" -> MobileRoute.Vendor.Summary
            route == "vendor/statement" -> MobileRoute.Vendor.Statement
            route == "vendor/payout" -> MobileRoute.Vendor.Payout
            route == "vendor/transaction" -> MobileRoute.Vendor.Transaction
            route == "vendor/attachment" -> MobileRoute.Vendor.Attachment
            route == "catalog" -> MobileRoute.Catalog.Root
            route == "catalog/browse" -> MobileRoute.Catalog.Browse
            route == "catalog/search" -> MobileRoute.Catalog.Search(query["q"] ?: query["searchText"])
            segments.size == 3 && segments[0] == "catalog" && segments[1] == "node" -> MobileRoute.Catalog.Node(segments[2])
            route == "message" -> MobileRoute.Message.Inbox
            segments.size == 3 && segments[0] == "message" && segments[1] == "thread" -> MobileRoute.Message.Thread(segments[2])
            route == "notification" -> MobileRoute.Notification
            route == "support" -> MobileRoute.Support.Home
            route == "support/case" -> MobileRoute.Support.Cases
            segments.size == 3 && segments[0] == "support" && segments[1] == "case" -> MobileRoute.Support.CaseDetail(segments[2])
            route.startsWith("support/") -> MobileRoute.Support.Flow(route)
            route == "attachment" -> MobileRoute.Attachment.Root
            segments.size == 2 && segments[0] == "attachment" -> MobileRoute.Attachment.Detail(segments[1])
            route == "cart" -> MobileRoute.Cart.Current
            route == "cart/checkout" -> MobileRoute.Cart.Checkout
            segments.size == 4 && segments[0] == "cart" && segments[1] == "checkout" && segments[2] == "result" -> MobileRoute.Cart.CheckoutResult(segments[3])
            route == "vendor/retail" -> MobileRoute.Product.Listing
            route == "vendor/retail/new" -> MobileRoute.Product.New
            segments.size == 4 && segments[0] == "vendor" && segments[1] == "retail" && segments[3] == "placement" -> MobileRoute.Product.Placement(segments[2])
            segments.size == 3 && segments[0] == "vendor" && segments[1] == "product" -> MobileRoute.Product.Detail(segments[2])
            route == "vendor/order" -> MobileRoute.Order.Listing
            route == "vendor/order/new" -> MobileRoute.Order.New
            segments.size == 3 && segments[0] == "vendor" && segments[1] == "order" -> MobileRoute.Order.Detail(segments[2])
            segments.size == 4 && segments[0] == "vendor" && segments[1] == "order" && segments[3] == "shipment" -> MobileRoute.Order.Shipment(segments[2])
            segments.size == 5 && segments[0] == "vendor" && segments[1] == "order" && segments[3] == "shipment" -> MobileRoute.Order.ShipmentDetail(segments[2], segments[4])
            segments.size == 4 && segments[0] == "vendor" && segments[1] == "order" && segments[3] == "tax" -> MobileRoute.Order.Tax(segments[2])
            segments.size == 5 && segments[0] == "vendor" && segments[1] == "order" && segments[3] == "tax" -> MobileRoute.Order.TaxDetail(segments[2], segments[4])
            route == "vendor/project" -> MobileRoute.Project.Listing
            route == "vendor/project/new" -> MobileRoute.Project.New
            segments.size == 3 && segments[0] == "vendor" && segments[1] == "project" -> MobileRoute.Project.Detail(segments[2])
            else -> null
        }
    }

    fun resolve(link: MobileLink): MobileRoute? = resolve(link.route, link.query)

    fun isKnownRoute(rawRoute: String?): Boolean = resolve(rawRoute) != null

    fun isSignOutAction(action: String?, rawRoute: String?): Boolean =
        action == "access.sign_out" || resolve(rawRoute) == MobileRoute.Access.SignOut

    fun isCurrentlyRenderable(rawRoute: String?): Boolean = when (resolve(rawRoute)) {
        MobileRoute.Dashboard,
        MobileRoute.More,
        MobileRoute.Money,
        MobileRoute.Wallet.Overview,
        MobileRoute.Wallet.Transaction,
        MobileRoute.Wallet.Funding,
        MobileRoute.Wallet.Withdrawal,
        is MobileRoute.Wallet.WithdrawalDetail,
        MobileRoute.Cart.Current,
        MobileRoute.Vendor.Overview,
        MobileRoute.Vendor.Profile,
        MobileRoute.Vendor.Summary,
        MobileRoute.Vendor.Statement,
        MobileRoute.Vendor.Payout,
        MobileRoute.Vendor.Transaction,
        MobileRoute.Vendor.Attachment,
        MobileRoute.Message.Inbox,
        MobileRoute.Notification,
        MobileRoute.Support.Home,
        MobileRoute.Support.Cases,
        is MobileRoute.Support.CaseDetail,
        is MobileRoute.Support.Flow,
        MobileRoute.Product.Listing,
        MobileRoute.Product.New,
        is MobileRoute.Product.Placement,
        is MobileRoute.Product.Detail,
        MobileRoute.Order.Listing,
        MobileRoute.Order.New,
        is MobileRoute.Order.Detail,
        is MobileRoute.Order.Shipment,
        is MobileRoute.Order.ShipmentDetail,
        is MobileRoute.Order.Tax,
        is MobileRoute.Order.TaxDetail,
        MobileRoute.Project.Listing,
        MobileRoute.Project.New,
        is MobileRoute.Project.Detail,
        MobileRoute.Catalog.Root,
        MobileRoute.Attachment.Root -> true
        else -> false
    }

    fun normalizeRoute(rawRoute: String?): String {
        val normalized = rawRoute
            ?.trim()
            .orEmpty()
            .trim('/')
            .replace(Regex("/{2,}"), "/")

        return if (normalized == "vendor/profile") "vendor/page" else normalized
    }
}
